File: 784/letter-case-permutation.py
class Solution:
    def letterCasePermutation(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        if len(S) == 0:
            return [S]
        if len(S) == 1:
            if S[0] >= '0' and S[0] <= '9':
                return [S]
            return [S.lower(), S.upper()]

        tails = self.letterCasePermutation(S[1:])
        res = []
        leading = []
        if S[0] >= '0' and S[0] <= '9':
            leading.append(S[:1])
        else:
            leading.append(S[:1].lower())
            leading.append(S[:1].upper())
        
        for i in range(len(leading)):
            for j in range(len(tails)):
                res.append(leading[i]+tails[j])

        return res


S = "a1b2"
print(Solution().letterCasePermutation(S))

File: 784/letter-case-permutation.js
/**
 * @param {string} S
 * @return {string[]}
 */
var letterCasePermutation = function (S) {
    let res = [];
    if (S.length == 0) {
        res.push(S);
        return res;
    }
    if (S.length == 1) {
        if (S[0] >= '0' && S[0] <= '9')
            res.push(S);
        res.push(S.toLowerCase(), S.toUpperCase());
        return res;
    }

    let tails = letterCasePermutation(S.slice(1));
    let heads = [];
    if (S[0] >= '0' && S[0] <= '9') {
        heads.push(S[0]);
    } else {
        heads.push(S[0].toLowerCase(), S[0].toUpperCase());
    }

    for (let head of heads) {
        for (let tail of tails) {
            res.push(head + tail);
        }
    }

    // for (let i = 0; i < heads.length; i++) {
    //     for (let j = 0; j < tails.length; j++) {
    //         res.push(heads[i] + tails[j]);
    //     }
    // }
    return res;
};


S = "a1b2";
console.log(letterCasePermutation(S));
